be/src/star_cache/star_cache.cpp

```diff
@@ -0,0 +1,395 @@
+    ac_config.flush_probability = config::star_cache_admission_flush_probability;
+    ac_config.delete_probability = config::star_cache_admission_delete_probability;
+    _admission_policy = new SizeBasedAdmissionPolicy(ac_config);
+
+    CapacityBasedPromotionPolicy::Config pm_config;
+    pm_config.mem_cap_threshold = config::star_cache_promotion_mem_threshold;
+    _promotion_policy = new CapacityBasedPromotionPolicy(pm_config);
+
+    return Status::OK();
+}
+
+Status StarCache::set(const std::string& cache_key, const IOBuf& buf, uint64_t ttl_seconds) {
+    if (buf.empty()) {
+        return Status::InternalError("cache value should not be empty");
+    }
+
+    CacheId cache_id = cachekey2id(cache_key);
+    CacheItemPtr cache_item = _access_index->find(cache_id);
+    if (cache_item) {
+        // TODO: Replace the target data directly.
+        _remove_cache_item(cache_id, cache_item);
+    }
+
+    size_t block_size = config::star_cache_block_size;
+    size_t block_count = (buf.size()  - 1) / block_size + 1;
+    uint64_t expire_time = MonotonicSeconds() + ttl_seconds;
+    cache_item = std::make_shared<CacheItem>(cache_key, block_count, buf.size(), expire_time);
+
+    uint32_t start_block_index = 0;
+    uint32_t end_block_index = (buf.size() - 1) / block_size;
+    for (uint32_t index = start_block_index; index <= end_block_index; ++index) {
+        BlockKey block_key = { .cache_id = cache_id, .block_index = index };
+        off_t offset = index * block_size;
+        IOBuf seg_buf;
+        if (block_count > 1) {
+            buf.append_to(&seg_buf, block_size, offset);
+        } else {
+            seg_buf = buf;
+        }
+        RETURN_IF_ERROR(_write_block(cache_item, block_key, seg_buf));
+    }
+
+    _access_index->insert(cache_id, cache_item);
+    return Status::OK();
+}
+
+Status StarCache::_write_block(CacheItemPtr cache_item, const BlockKey& block_key, const IOBuf& buf) {
+    BlockItem& block = cache_item->blocks[block_key.block_index];
+    auto location = _promotion_policy->check_write(cache_item, block_key);
+
+    if (location == BlockLocation::MEM) {
+        // allocate segment and evict
+        auto segment = _alloc_block_segment(block_key, 0, buf);
+        if (!segment) {
+            return Status::InternalError("allocate segment failed");
+        }
+        auto mem_block = _mem_cache->new_block_item(block_key, BlockState::DIRTY);
+        Status st = _mem_cache->write_block(block_key, mem_block, { segment });
+        if(!st.ok() || !_set_mem_block(cache_item, block_key.block_index, mem_block)) {
+            _mem_cache->free_block_segment(segment);
+            delete mem_block;
+            return Status::InternalError("the block is released");
+        }
+        _mem_cache->evict_track(block_key);
+
+    } else if (location == BlockLocation::DISK) {
+        auto disk_block = block.disk_block_item;
+        if (disk_block) {
```

Comment:
when is this pointer not null

Reply:
> when is this pointer not null

 If we replace the original cache item content, it is may be not null. But now it is impossible, I'll remove this branch.